# APLICACIONES EN CIENCIAS DE COMPUTACION

## Laboratorio 2: Implementacion metodos de busqueda y heurísticas  para el Problema de Busqueda de Rutas en Mapas

La tarea de este laboratorio consiste en implementar y comparar métodos de busqueda ciega y funciones heurísticas para A* para buscar rutas en mapas. La clase MapSearchProblem necesita ser completada (metodos actions y path_cost). Deberá implementarse 3 heurísticas para A* (h1:distancia en linea recta, h2:distancia Manhatan y 3:distancia Chebyshev). Abajo puede encontrar la definición de la clase Mapa y una instancia específica con el mapa de Romania que sera usado en las pruebas. Mapa tiene dos atributos: <b>neighbors</b> (diccionario que contiene las ciudades vecinas de cada ciudad y las distancias para llegar a ellas) y <b>location</b>, diccionario con las coordenadas X,Y de cada ciudad. 

Al final de este notebook debe responder algunas preguntas . 


<b>Mapa de Romania para realizar pruebas</b>

In [1]:
class Mapa:
    neighbors = dict()
    location = dict()
    
romania = Mapa()
romania.neighbors = {
 'A': [('Z',75), ('T',118), ('S',140)],
 'B': [('F',211), ('P',101), ('G',90), ('U',85)],
 'C': [('D',120), ('R',146), ('P',138)],
 'D': [('M',75), ('C',120)],
 'E': [('H',86)],
 'F': [('S',99), ('B',211)],
 'G': [('B',90)],
 'H': [('U',98), ('E',86)],
 'I': [('N',87), ('V',92)],
 'L': [('T',111), ('M',70)],
 'M': [('L',70), ('D',75)],
 'N': [('I',87)],
 'O': [('Z',71), ('S',151)],
 'P': [('R',97), ('C',138), ('B',101)],
 'R': [('S',80), ('C',146), ('P',97)],
 'S': [('A',140), ('O',151), ('F',99), ('R',80)],
 'T': [('A',118), ('L',111)],
 'U': [('B',85), ('V',142), ('H',98)],
 'V': [('U',142), ('I',92)],
 'Z': [('O',71), ('A',75)]}
romania.location = {
 'A': (91, 492),
 'B': (400, 327),
 'C': (253, 288),
 'D': (165, 299),
 'E': (562, 293),
 'F': (305, 449),
 'G': (375, 270),
 'H': (534, 350),
 'I': (473, 506),
 'L': (165, 379),
 'M': (168, 339),
 'N': (406, 537),
 'O': (131, 571),
 'P': (320, 368),
 'R': (233, 410),
 'S': (207, 457),
 'T': (94, 410),
 'U': (456, 350),
 'V': (509, 444),
 'Z': (108, 531)}

### Clase <b>SearchProblem</b>

Esta es una clase abstracta para definir problemas de busqueda. Se debe hacer subclases que implementen los metodos de las acciones, resultados, test de objetivo y el costo de camino. Entonces se puede instanciar las subclases y resolverlos con varias funciones de busqueda.

In [16]:
class SearchProblem(object):
    def __init__(self, initial, goal=None):
        """Este constructor especifica el estado inicial y posiblemente el estado(s) objetivo(s),
        La subclase puede añadir mas argumentos."""
        self.initial = initial
        self.goal = goal

    def actions(self, state):
        """Retorna las acciones que pueden ser ejecutadas en el estado dado.
        El resultado es tipicamente una lista."""
        raise NotImplementedError

    def result(self, state, action):
        """Retorna el estado que resulta de ejecutar la accion dada en el estado state.
        La accion debe ser alguna de self.actions(state)."""
        raise NotImplementedError

    def goal_test(self, state):
        """Retorna True si el estado pasado satisface el objetivo."""
        raise NotImplementedError

    def path_cost(self, c, state1, action, state2):
        """Retorna el costo del camino de state2 viniendo de state1 con 
        la accion action, asumiendo un costo c para llegar hasta state1. 
        El metodo por defecto cuesta 1 para cada paso en el camino."""
        return c + 1

    def value(self, state):
        """En problemas de optimizacion, cada estado tiene un valor. Algoritmos
        como Hill-climbing intentan maximizar este valor."""
        raise NotImplementedError

###  <b> Clase MapSearchProblem </b>  
Esta es una subclase de SearchProblem donde se define concretamente el problema de busqueda en mapa. El constructor recibe el estado inicial, objetivo y un mapa. Se necesita completar Actions (acciones disponibles para un estado dado) y path_cost.

In [33]:
class MapSearchProblem(SearchProblem):
    def __init__(self, initial, goal, mapa):
        """El constructor recibe  el estado inicial, el estado objetivo y un mapa (de clase Mapa)"""
        self.initial = initial
        self.goal = goal
        self.map = mapa

    def actions(self, state):
        """Retorna las acciones ejecutables desde ciudad state.
        El resultado es una lista de strings tipo 'goCity'. 
        Por ejemplo, en el mapa de Romania, las acciones desde Arad serian:
         ['goZerind', 'goTimisoara', 'goSibiu']"""
        ## TODO
        tuplas = self.map.neighbors[state]
        salidas = list(tuplas)
        for i in range(0,len(tuplas)):
            salidas[i] = 'go' + tuplas[i][0]
        return salidas
        
    def result(self, state, action):
        """Retorna el estado que resulta de ejecutar la accion dada desde ciudad state.
        La accion debe ser alguna de self.actions(state)
        Por ejemplo, en el mapa de Romania, el resultado de aplicar la accion 'goZerind' 
        desde el estado 'Arad' seria 'Zerind'"""  
        newState = action[2]
        return newState
        
    def goal_test(self, state):
        """Retorna True si state es self.goal"""
        return (self.goal == state) 

    def path_cost(self, c, state1, action, state2):
        """Retorna el costo del camino de state2 viniendo de state1 con la accion action 
        El costo del camino para llegar a state1 es c. El costo de la accion debe ser
        extraido de self.map."""
        ## TODO
        tuplas = self.map.neighbors[state1]
        # se buscara el estado siguiente con el state2
        for tupla in tuplas:
            if(tupla[0]==state2):
                distancia = tupla[1]
                break
        return c + distancia
        

### Clase <b>Node</b>

Estructura de datos para almacenar la informacion de un nodo en un <b>arbol de busqueda</b>. Contiene información del nodo padre y el estado que representa el nodo. Tambien incluye la accion que nos llevo al presente nodo y el costo total del camino desde el nodo raiz hasta este nodo.

In [18]:
class Node:
    def __init__(self, state, parent=None, action=None, path_cost=0):
        "Crea un nodo de arbol de busqueda, derivado del nodo parent y accion action"
        self.state = state
        self.parent = parent
        self.action = action
        self.path_cost = path_cost
        self.depth = 0
        if parent:
            self.depth = parent.depth + 1

    def expand(self, problem):
        "Devuelve los nodos alcanzables en un paso a partir de este nodo."
        return [self.child_node(problem, action)
                for action in problem.actions(self.state)]

    def child_node(self, problem, action):
        next = problem.result(self.state, action)
        return Node(next, self, action,
                    problem.path_cost(self.path_cost, self.state, action, next))

    def solution(self):
        "Retorna la secuencia de acciones para ir de la raiz a este nodo."
        return [node.action for node in self.path()[1:]]

    def path(self):
        "Retorna una lista de nodos formando un camino de la raiz a este nodo."
        node, path_back = self, []
        while node:
            path_back.append(node)
            node = node.parent
        return list(reversed(path_back))
    
    def __eq__(self, other): 
        "Este metodo se ejecuta cuando se compara nodos. Devuelve True cuando los estados son iguales"
        return isinstance(other, Node) and self.state == other.state

### <b> Frontera tipo cola FIFO (first-in first out) para BFS</b> 

In [19]:
from collections import deque

class FIFOQueue(deque):
    """Una cola First-In-First-Out"""
    def pop(self):
        return self.popleft()

### <b> Frontera tipo cola de prioridad ordenada por una funcion de costo (para best_first_graph_search y A*)</b> 

In [20]:
import heapq
class FrontierPQ:
    "Una Frontera ordenada por una funcion de costo (Priority Queue)"
    
    def __init__(self, initial, costfn=lambda node: node.path_cost):
        "Inicializa la Frontera con un nodo inicial y una funcion de costo especificada (por defecto es el costo de camino)."
        self.heap   = []
        self.states = {}
        self.costfn = costfn
        self.add(initial)
    
    def add(self, node):
        "Agrega un nodo a la frontera."
        cost = self.costfn(node)
        heapq.heappush(self.heap, (cost, node))
        self.states[node.state] = node
        
    def pop(self):
        "Remueve y retorna el nodo con minimo costo."
        (cost, node) = heapq.heappop(self.heap)
        self.states.pop(node.state, None) # remove state
        return node
    
    def replace(self, node):
        "node reemplaza al nodo de la Fontera que tiene el mismo estado que node."
        if node.state not in self:
            raise ValueError('{} no tiene nada que reemplazar'.format(node.state))
        for (i, (cost, old_node)) in enumerate(self.heap):
            if old_node.state == node.state:
                self.heap[i] = (self.costfn(node), node)
                heapq._siftdown(self.heap, 0, i)
                return

    def __contains__(self, state): return state in self.states
    
    def __len__(self): return len(self.heap)

### <b>Algoritmo general de búsqueda con memoria de nodos expandidos (Graph Search)</b>

Algoritmo de general de busqueda ciega con memoria de estados visitados. El argumento frontier debe ser una cola vacia. Si la frontera es tipo FIFO hace busqueda en amplitud (BFS), si la frontera es una pila hará busqueda en profundidad (DFS)

In [21]:
def graph_search(problem, frontier):
    frontier.append(Node(problem.initial))
    explored = set()     # memoria de estados visitados
    expanded_nodes = 0   # contador de nodos expandidos
    while frontier:
        node = frontier.pop()
        if problem.goal_test(node.state):
            return node, expanded_nodes
        explored.add(node.state)
        expanded_nodes = expanded_nodes + 1
        frontier.extend(child for child in node.expand(problem)
                        if child.state not in explored and
                        child not in frontier)
    return None

### <b> Algoritmo Best-First-Graph-Search </b> 
Algoritmo general de busqueda con información. La frontera es una cola de prioridad ordenada por la funcion de evaluacion f 

In [22]:
def best_first_graph_search(problem, f):
    """Busca el objetivo expandiendo el nodo de la frontera con el menor valor de la funcion f. Memoriza estados visitados
    Antes de llamar a este algoritmo hay que especificar La funcion f(node). Si f es node.depth tenemos Busqueda en Amplitud; 
    si f es node.path_cost tenemos Busqueda  de Costo Uniforme. Si f es una heurística tenemos Busqueda Voraz;
    Si f es node.path_cost + heuristica(node) tenemos A* """

    frontier = FrontierPQ( Node(problem.initial), f )  # frontera tipo cola de prioridad ordenada por f
    explored = set()     # memoria de estados visitados
    expanded_nodes = 0   # contador de nodos expandidos
    while frontier:
        node = frontier.pop()
        if problem.goal_test(node.state):
            return node, expanded_nodes
        explored.add(node.state)
        expanded_nodes = expanded_nodes + 1
        for action in problem.actions(node.state):
            child = node.child_node(problem, action)
            if child.state not in explored and child.state not in frontier:
                frontier.add(child)
            elif child.state in frontier:
                incumbent = frontier.states[child.state] 
                if f(child) < f(incumbent):
                    frontier.replace(child)

### <b> Algoritmo A* </b> 
A* es un caso especial de best_first_graph_search con f = path_cost + heuristic

In [23]:
def astar_search(problem, heuristic):
    f = lambda node: node.path_cost + heuristic(node, problem)
    return best_first_graph_search(problem, f)

def nullheuristic(node, problem):   # heurística nula (A* se convierte en busqueda de costo uniforme)
    return 0

### <b> Heurísticas para A* </b> 
Se debe implementar las heurísticas abajo para A* 

In [48]:
import math

def distance(estadoInicial,solution,problem):
        "Retorna la distancia desde la raiz a este nodo"
        estado = estadoInicial
        estadoSig = ''
        recorrido = 0
        distancias = problem.map.neighbors
        for accion in solution:
            estadoSig = accion[2]
            opciones = distancias[estado]
            for i in opciones:
                if i[0] == estadoSig:
                    recorrido = recorrido + i[1]
                    break
            estado = estadoSig
        return recorrido
            
def h1(node, problem):
    "Distancia en linea recta desde la posicion de node hasta la posicion del Objetivo (problem.goal)"
    ## TODO
    
    coord1 = problem.map.location[node.state]
    coord2 = problem.map.location[problem.goal]
    dx = coord2[0]-coord1[0]
    dy = coord2[1]-coord1[1]
    return math.sqrt(pow(dx,2) + pow(dy,2))

def h2(node, problem):
    "Distancia Manhatan (o city block) desde la posicion de node hasta la posicion del Objetivo (problem.goal)"
    ## TODO
    coord1 = problem.map.location[node.state]
    coord2 = problem.map.location[problem.goal]
    dx = coord2[0]-coord1[0]
    dy = coord2[1]-coord1[1]
    return abs(dx) + abs(dy)

def h3(node, problem):
    "Distancia Chebyshev (el mayor de los catetos) desde la posicion de node hasta la posicion del Objetivo (problem.goal)"
    ## TODO
    coord1 = problem.map.location[node.state]
    coord2 = problem.map.location[problem.goal]
    dx = coord2[0]-coord1[0]
    dy = coord2[1]-coord1[1]
    return max(abs(dx),abs(dy))


### <b> Probando los algoritmos de Busqueda</b> 


In [54]:
estadoInicial = 'A'
estadoFinal = 'B'
p = MapSearchProblem(estadoInicial, estadoFinal, romania)   # problema de busqueda de ruta de Arad a Bucharest

node, num_exp_nodes = graph_search(p, FIFOQueue())
print( 'Solucion obtenida con BFS: {}. Nodos expandidos = {}. Costo de la solucion = {}'.format(node.solution(), num_exp_nodes,distance(estadoInicial,node.solution(),p)) )

node, num_exp_nodes = graph_search(p, [])
print( 'Solucion obtenida con DFS: {}. Nodos expandidos = {}. Costo de la solucion = {}'.format(node.solution(), num_exp_nodes,distance(estadoInicial,node.solution(),p)) )

node, num_exp_nodes = astar_search(p,nullheuristic)
print( 'Solucion obtenida con A* y heuristica nula (UCS): {}. Nodos expandidos = {}. Costo de la solucion = {}'.format(node.solution(), num_exp_nodes,distance(estadoInicial,node.solution(),p)) )

node, num_exp_nodes = astar_search(p, h1)
print( 'Solucion obtenida con A* y heuristica h1: {}. Nodos expandidos = {}. Costo de la solucion = {}'.format(node.solution(), num_exp_nodes,distance(estadoInicial,node.solution(),p)) )

node, num_exp_nodes = astar_search(p, h2)
print( 'Solucion obtenida con A* y heuristica h2: {}. Nodos expandidos = {}. Costo de la solucion = {}'.format(node.solution(), num_exp_nodes,distance(estadoInicial,node.solution(),p)) )

node, num_exp_nodes = astar_search(p, h3)
print( 'Solucion obtenida con A* y heuristica h3: {}. Nodos expandidos = {}. Costo de la solucion = {}'.format(node.solution(), num_exp_nodes,distance(estadoInicial,node.solution(),p)) )


Solucion obtenida con BFS: ['goS', 'goF', 'goB']. Nodos expandidos = 9. Costo de la solucion = 450
Solucion obtenida con DFS: ['goS', 'goR', 'goP', 'goB']. Nodos expandidos = 4. Costo de la solucion = 418
Solucion obtenida con A* y heuristica nula (UCS): ['goS', 'goR', 'goP', 'goB']. Nodos expandidos = 12. Costo de la solucion = 418
Solucion obtenida con A* y heuristica h1: ['goS', 'goR', 'goP', 'goB']. Nodos expandidos = 5. Costo de la solucion = 418
Solucion obtenida con A* y heuristica h2: ['goS', 'goF', 'goB']. Nodos expandidos = 3. Costo de la solucion = 450
Solucion obtenida con A* y heuristica h3: ['goS', 'goR', 'goP', 'goB']. Nodos expandidos = 7. Costo de la solucion = 418


## Preguntas

<b>1) Probar BFS, DFS, A* con heuristicas: nula, h1, h2, y h3 en los siguientes problemas (registre los nodos expandidos y los costos de las soluciones): </b>
       - p = MapSearchProblem('A', 'B', romania)
        Solucion obtenida con BFS: ['goS', 'goF', 'goB']. Nodos expandidos = 9. Costo de la solucion = 450
        Solucion obtenida con DFS: ['goS', 'goR', 'goP', 'goB']. Nodos expandidos = 4. Costo de la solucion = 418
        Solucion obtenida con A* y heuristica nula (UCS): ['goS', 'goR', 'goP', 'goB']. Nodos expandidos = 12. Costo de la solucion = 418
        Solucion obtenida con A* y heuristica h1: ['goS', 'goR', 'goP', 'goB']. Nodos expandidos = 5. Costo de la solucion = 418
        Solucion obtenida con A* y heuristica h2: ['goS', 'goF', 'goB']. Nodos expandidos = 3. Costo de la solucion = 450
        Solucion obtenida con A* y heuristica h3: ['goS', 'goR', 'goP', 'goB']. Nodos expandidos = 7. Costo de la solucion = 418

       - p = MapSearchProblem('E', 'O', romania)
        Solucion obtenida con BFS: ['goH', 'goU', 'goB', 'goF', 'goS', 'goO']. Nodos expandidos = 14. Costo de la solucion = 730
        Solucion obtenida con DFS: ['goH', 'goU', 'goB', 'goP', 'goC', 'goD', 'goM', 'goL', 'goT', 'goA', 'goS', 'goO']. Nodos expandidos = 16. Costo de la solucion = 1293
        Solucion obtenida con A* y heuristica nula (UCS): ['goH', 'goU', 'goB', 'goP', 'goR', 'goS', 'goO']. Nodos expandidos = 15. Costo de la solucion = 698
        Solucion obtenida con A* y heuristica h1: ['goH', 'goU', 'goB', 'goP', 'goR', 'goS', 'goO']. Nodos expandidos = 8. Costo de la solucion = 698
        Solucion obtenida con A* y heuristica h2: ['goH', 'goU', 'goB', 'goP', 'goR', 'goS', 'goO']. Nodos expandidos = 7. Costo de la solucion = 698
        Solucion obtenida con A* y heuristica h3: ['goH', 'goU', 'goB', 'goP', 'goR', 'goS', 'goO']. Nodos expandidos = 9. Costo de la solucion = 698
        
       - p = MapSearchProblem('U', 'O', romania)
        Solucion obtenida con BFS: ['goB', 'goF', 'goS', 'goO']. Nodos expandidos = 14. Costo de la solucion = 546
        Solucion obtenida con DFS: ['goB', 'goP', 'goC', 'goD', 'goM', 'goL', 'goT', 'goA', 'goS', 'goO']. Nodos expandidos = 16. Costo de la solucion = 1109
        Solucion obtenida con A* y heuristica nula (UCS): ['goB', 'goP', 'goR', 'goS', 'goO']. Nodos expandidos = 15. Costo de la solucion = 514
        Solucion obtenida con A* y heuristica h1: ['goB', 'goP', 'goR', 'goS', 'goO']. Nodos expandidos = 6. Costo de la solucion = 514
        Solucion obtenida con A* y heuristica h2: ['goB', 'goP', 'goR', 'goS', 'goO']. Nodos expandidos = 5. Costo de la solucion = 514
        Solucion obtenida con A* y heuristica h3: ['goB', 'goP', 'goR', 'goS', 'goO']. Nodos expandidos = 8. Costo de la solucion = 514
  
<b>2) Compare los algoritmos de busqueda ciega (BFS, DFS y UCS) en cuanto a sus costos temporales y optimalidad </b>
      Se observa que en los tres casos, si bien la tendencia no es notoria, es UCS el que toma o la mayor cantidad de nodos expandidos o una cantidad muy cercana al máximo.
      En cuanto a BFS y DFS depende mucho de la configuración del mapa para ver su comportamiento. Curiosamente, las cantidades de nodos expandidos en las rutas 'E'-'O' y 'U'-'O' son las mismas, lo que podría hacer sospechar una especie de simetría en estas rutas.
      En los tres casos UCS encuentra el resultado óptimo. Por otro lado, tanto el uso de BFS como el uso de DFS no nos garantizan optimalidad.

<b>3) Todas las heurísticas encuentran la solución optima? Por Que? </b>
      No, porque solo optimizan su forma de ver el recorrido. Para las diferentes heurísticas hay resultados distintos, se enfocan en su forma de solucionarlo. Esto se pudo observar en la ruta 'A'-'B' donde usando la heurística 2 con A* no se encontró la solución óptima
<b>4) Cual heurística expande menos nodos? Por Que? </b>
      A* y heuristica h2.